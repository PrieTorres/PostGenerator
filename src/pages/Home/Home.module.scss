.container {
  min-height: 100vh;
  background-color: #222;
  color: #fff;
  text-align: center;
  padding-bottom: 20px;
  overflow-x: hidden;
}
