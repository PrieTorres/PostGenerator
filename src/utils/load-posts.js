export const loadPosts = async (page = 0, postPerPage = 8) => {
  const postsFetch = fetch("https://jsonplaceholder.typicode.com/posts")
  const imagesFetch = fetch(`https://api.pexels.com/v1/search/?page=${page}&per_page=${postPerPage}&query=dog`, {
    headers: {
      Authorization: "RZUilClUAopA89M1o2WlLhrzdMdJpbKDPzfWb0AW9IS9hBGyicS8m1kr"
    }
  })

  const [posts, images] = await Promise.all([postsFetch, imagesFetch]);

  const postJson = await posts.json();
  const imagesJson = await images.json();

  const photos = imagesJson.photos;

  const postsAndPhotos = photos.map((photo, i) => {
    return { ...postJson[i], photo }
  });

  return {posts: postsAndPhotos, postQuantity: parseInt(postJson.length)};
}