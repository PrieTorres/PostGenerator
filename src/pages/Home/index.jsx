import { useEffect, useState } from "react";
import style from "./Home.module.scss";
import { Posts } from "../../components/Posts";
import { loadPosts } from "../../utils/load-posts";
import { Button } from "../../components/Button";
import { TextFieldSearch } from "../../components/TextFieldSearch";
import P from "prop-types";


export const Home = ({ testPosts = [] }) => {
  // posts as parameter just for tests
  const [posts, setPosts] = useState([...testPosts]);
  const [page, setPage] = useState(1);
  const [postQuantity, setPostQuantity] = useState(0);
  const [searchActive, setSearchActive] = useState(false);
  const [searchValue, setSearchValue] = useState("");
  const [correspondedPosts, setCorrespondedPosts] = useState([]);
  const postsPerPage = 8;
  const noData = page * postsPerPage >= postQuantity;

  useEffect(() => {
    if (testPosts.length <= 0) {
      handleLoadPosts();
    }
  }, []);

  useEffect(() => {
    if (testPosts.length <= 0) {
      handleLoadPosts();
    }
    scrollToTop();
  }, [page]);

  const handleLoadPosts = async () => {
    if (testPosts.length > 0) {
      return;
    }
    const postsAndPhotos = await loadPosts(page, postsPerPage);

    let updatedPostList = [...postsAndPhotos.posts, ...posts];

    if (postsAndPhotos.postQuantity > postQuantity) {
      setPostQuantity(postsAndPhotos.postQuantity);
    }
    setPosts(updatedPostList);
  };

  const loadMorePosts = () => {
    let nextPage = page + 1;

    setPage(nextPage);
  };

  const scrollToTop = () => {
    window.scrollTo({ top: 0, behavior: "smooth" });
  };

  const handleSearch = (e) => {
    const { value } = e.target;
    let responseSearch = [];

    posts.forEach((post, i) => {
      if (
        (post.body && post.body.includes(value)) ||
        (post.title && post.title.includes(value))
      ) {
        posts[i].inSearch = true;
      } else {
        posts[i].inSearch = false;
      }
    });

    responseSearch = posts.filter((post) => post.inSearch);

    setSearchValue(value);
    setCorrespondedPosts(responseSearch);
    setSearchActive(true);
  };

  return (
    <div className={style.container} data-testid="home-page">
      <TextFieldSearch value={searchValue} onChange={handleSearch} />
      <Posts posts={searchActive ? correspondedPosts : posts} />
      <Button
        disabled={noData}
        label={"carregar mais posts..."}
        onClick={loadMorePosts}
        extraStyles={{ marginTop: "10px" }}
      />
    </div>
  );
};

export default Home;

Home.defaultProps = {
  testPosts: []
}

Home.propTypes = {
  testPosts: P.array
}
