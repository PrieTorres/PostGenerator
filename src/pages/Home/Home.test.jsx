import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import Home from ".";
import userEvent from "@testing-library/user-event";
import { loadPosts } from "../../utils/load-posts";


describe('<Home />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  })
  afterAll(() => {
    jest.clearAllMocks();
  })

  window.scrollTo = jest.fn();
  test("render home page", async () => {
    const { container } = await render(<Home />);
    const page = screen.getByTestId("home-page");

    await waitFor(() => {
      expect(page).toBeInTheDocument();
      expect(container.firstChild).toMatchSnapshot(); // vai criar a pasta snapshot ;))
    })
  });

  test("handle search", async () => {
    const { container } = await render(<Home />);
    const input = container.querySelector("input");
    // const posts = await screen.findAllByTestId("loaded-post");
    userEvent.type(input, "quit");
    const matchedPosts = await screen.findAllByTestId("loaded-post");

    await waitFor(() => {
      expect(input).toHaveValue('quit');
      expect(matchedPosts).toHaveLength(2);
    })
  });


})
