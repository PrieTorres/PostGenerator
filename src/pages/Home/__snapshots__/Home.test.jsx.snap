// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> render home page 1`] = `
<div
  class="container"
  data-testid="home-page"
>
  <div
    class="container"
  >
    <input
      class="input"
      placeholder="burcar..."
      type="search"
    />
  </div>
  <div
    class="posts"
  >
    <div
      class="no-posts"
    >
      No posts... ;(
    </div>
  </div>
  <div
    class="button-container"
    style="margin-top: 10px;"
  >
    <button
      class="button"
      disabled=""
    >
      carregar mais posts...
    </button>
  </div>
</div>
`;
